File: app/src/main/java/social/entourage/android/view/CountryCodePicker/CountryCodeDialog.java
package social.entourage.android.view.countrycodepicker;

import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Bundle;
import androidx.appcompat.widget.AppCompatEditText;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.Window;
import android.view.inputmethod.InputMethodManager;
import android.widget.RelativeLayout;

import java.util.ArrayList;
import java.util.List;

import social.entourage.android.R;

/**
 * Dialog for selecting Country.
 */

class CountryCodeDialog extends Dialog {
    private AppCompatEditText mEdtSearch;
    private AppCompatTextView mTvNoResult;
    private AppCompatTextView mTvTitle;
    private RecyclerView mRvCountryDialog;
    private CountryCodePicker mCountryCodePicker;
    private RelativeLayout mRlyDialog;

    private List<Country> masterCountries;
    private List<Country> mFilteredCountries;
    private InputMethodManager mInputMethodManager;
    private CountryCodeAdapter mAdapter;
    private List<Country> mTempCountries;

    CountryCodeDialog(CountryCodePicker countryCodePicker) {
        super(countryCodePicker.getContext());
        this.mCountryCodePicker = countryCodePicker;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.layout_code_picker_dialog);
        setupUI();
        setupData();
    }

    private void setupUI() {
        mRlyDialog = this.findViewById(R.id.dialog_rly);
        mRvCountryDialog = this.findViewById(R.id.country_dialog_rv);
        mTvTitle = this.findViewById(R.id.title_tv);
        mEdtSearch = this.findViewById(R.id.search_edt);
        mTvNoResult = this.findViewById(R.id.no_result_tv);
    }

    private void setupData() {
        if (mCountryCodePicker.getTypeFace() != null) {
            Typeface typeface = mCountryCodePicker.getTypeFace();
            mTvTitle.setTypeface(typeface);
            mEdtSearch.setTypeface(typeface);
            mTvNoResult.setTypeface(typeface);
        }
        if (mCountryCodePicker.getBackgroundColor() != mCountryCodePicker.getDefaultBackgroundColor()) {
            mRlyDialog.setBackgroundColor(mCountryCodePicker.getBackgroundColor());
        }

        if (mCountryCodePicker.getTextColor() != mCountryCodePicker.getDefaultContentColor()) {
            int color = mCountryCodePicker.getTextColor();
            mTvTitle.setTextColor(color);
            mTvNoResult.setTextColor(color);
            mEdtSearch.setTextColor(color);
            mEdtSearch.setHintTextColor(adjustAlpha(color, 0.7f));
        }

        mCountryCodePicker.refreshCustomMasterList();
        mCountryCodePicker.refreshPreferredCountries();
        masterCountries = mCountryCodePicker.getCustomCountries(mCountryCodePicker);

        CountryCodeAdapter.Callback callback = new CountryCodeAdapter.Callback() {
            @Override
            public void onItemCountrySelected(Country country) {
                mCountryCodePicker.setSelectedCountry(country);
                //if (view != null && mCountries.get(position) != null) {
                mInputMethodManager.hideSoftInputFromWindow(mEdtSearch.getWindowToken(), 0);
                CountryCodeDialog.this.dismiss();
            }
        };

        this.mFilteredCountries = getFilteredCountries();

        mAdapter = new CountryCodeAdapter(mFilteredCountries, mCountryCodePicker, callback);
        if (!mCountryCodePicker.isSelectionDialogShowSearch()) {
            RelativeLayout.LayoutParams params =
                    (RelativeLayout.LayoutParams) mRvCountryDialog.getLayoutParams();
            params.height = RecyclerView.LayoutParams.WRAP_CONTENT;
            mRvCountryDialog.setLayoutParams(params);
        }
        mRvCountryDialog.setLayoutManager(new LinearLayoutManager(getContext()));
        mRvCountryDialog.setAdapter(mAdapter);

        mInputMethodManager = (InputMethodManager) mCountryCodePicker.getContext()
                .getSystemService(Context.INPUT_METHOD_SERVICE);
        setSearchBar();
    }

    void reShow() {
        setupData();
        show();
    }

    private int adjustAlpha(int color, float factor) {
        int alpha = Math.round(Color.alpha(color) * factor);
        int red = Color.red(color);
        int green = Color.green(color);
        int blue = Color.blue(color);
        return Color.argb(alpha, red, green, blue);
    }

    private void setSearchBar() {
        if (mCountryCodePicker.isSelectionDialogShowSearch()) {
            setTextWatcher();
        } else {
            mEdtSearch.setVisibility(View.GONE);
        }
    }

    /**
     * add textChangeListener, to apply new query each time editText get text changed.
     */
    private void setTextWatcher() {
        if (mEdtSearch != null) {
            mEdtSearch.addTextChangedListener(new TextWatcher() {

                @Override
                public void afterTextChanged(Editable s) {
                }

                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    applyQuery(s.toString());
                }
            });

            if (mCountryCodePicker.isKeyboardAutoPopOnSearch()) {
                if (mInputMethodManager != null) {
                    mInputMethodManager.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0);
                }
            }
        }
    }

    /**
     * Filter country list for given keyWord / query.
     * Lists all countries that contains @param query in country's name, name code or phone code.
     *
     * @param query : text to match against country name, name code or phone code
     */
    private void applyQuery(String query) {
        mTvNoResult.setVisibility(View.GONE);
        query = query.toLowerCase();

        //if query started from "+" ignore it
        if (query.length() > 0 && query.charAt(0) == '+') {
            query = query.substring(1);
        }

        mFilteredCountries = getFilteredCountries(query);

        if (mFilteredCountries.size() == 0) {
            mTvNoResult.setVisibility(View.VISIBLE);
        }

        mAdapter.notifyDataSetChanged();
    }

    private List<Country> getFilteredCountries() {
        return getFilteredCountries("");
    }

    private List<Country> getFilteredCountries(String query) {
        if (mTempCountries == null) {
            mTempCountries = new ArrayList<>();
        } else {
            mTempCountries.clear();
        }

        List<Country> preferredCountries = mCountryCodePicker.getPreferredCountries();
        if (preferredCountries != null && preferredCountries.size() > 0) {
            for (Country country : preferredCountries) {
                if (country.isEligibleForQuery(query)) {
                    mTempCountries.add(country);
                }
            }

            if (mTempCountries.size() > 0) { //means at least one preferred country is added.
                mTempCountries.add(null); // this will add separator for preference countries.
            }
        }

        for (Country country : masterCountries) {
            if (country.isEligibleForQuery(query)) {
                mTempCountries.add(country);
            }
        }
        return mTempCountries;
    }
}

File: app/src/main/java/social/entourage/android/tools/view/countrycodepicker/CountryCodeDialog.kt
package social.entourage.android.tools.view.countrycodepicker

import android.app.Dialog
import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.Window
import android.view.inputmethod.InputMethodManager
import android.widget.RelativeLayout
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import social.entourage.android.databinding.LayoutCodePickerDialogBinding
import social.entourage.android.tools.hideKeyboard
import java.util.Locale
import kotlin.math.roundToInt

/**
 * Dialog for selecting Country.
 */
internal class CountryCodeDialog(private val mCountryCodePicker: CountryCodePicker) : Dialog(mCountryCodePicker.context) {
    private lateinit var binding: LayoutCodePickerDialogBinding
    
    private var mFilteredCountries: List<Country?>? = null
    private var mInputMethodManager: InputMethodManager? = null
    private var mAdapter: CountryCodeAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        binding = LayoutCodePickerDialogBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setupData()
    }

    private fun setupData() {
        mCountryCodePicker.typeFace?.let { typeface->
            binding.titleTv.typeface = typeface
            binding.searchEdt.typeface = typeface
            binding.noResultTv.typeface = typeface
        }
        if (mCountryCodePicker.mBackgroundColor != CountryCodePicker.defaultBackgroundColor) {
            binding.dialogRly.setBackgroundColor(mCountryCodePicker.mBackgroundColor)
        }
        if (mCountryCodePicker.textColor != CountryCodePicker.defaultContentColor) {
            val color = mCountryCodePicker.textColor
            binding.titleTv.setTextColor(color)
            binding.noResultTv.setTextColor(color)
            binding.searchEdt.setTextColor(color)
            binding.searchEdt.setHintTextColor(adjustAlpha(color, 0.7f))
        }
        mCountryCodePicker.refreshCustomMasterList()
        mCountryCodePicker.refreshPreferredCountries()
        val callback: CountryCodeAdapter.Callback = object : CountryCodeAdapter.Callback {
            override fun onItemCountrySelected(country: Country?) {
                mCountryCodePicker.selectedCountry = country
                //if (view != null && mCountries.get(position) != null) {
                binding.searchEdt.hideKeyboard()
                dismiss()
            }
        }
        mFilteredCountries = getFilteredCountries("").also {
            mAdapter = CountryCodeAdapter(it, mCountryCodePicker, callback)
        }
        if (!mCountryCodePicker.isSelectionDialogShowSearch) {
            binding.countryDialogRv.let {
                val params = it.layoutParams as RelativeLayout.LayoutParams
                params.height = RecyclerView.LayoutParams.WRAP_CONTENT
                it.layoutParams = params
            }
        }
        binding.countryDialogRv.layoutManager = LinearLayoutManager(context)
        binding.countryDialogRv.adapter = mAdapter
        mInputMethodManager = mCountryCodePicker.context
                .getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        setSearchBar()
    }

    fun reShow() {
        setupData()
        show()
    }

    private fun adjustAlpha(color: Int, factor: Float): Int {
        val alpha = (Color.alpha(color) * factor).roundToInt()
        val red = Color.red(color)
        val green = Color.green(color)
        val blue = Color.blue(color)
        return Color.argb(alpha, red, green, blue)
    }

    private fun setSearchBar() {
        if (mCountryCodePicker.isSelectionDialogShowSearch) {
            setTextWatcher()
        } else {
            binding.searchEdt.visibility = View.GONE
        }
    }

    /**
     * add textChangeListener, to apply new query each time editText get text changed.
     */
    private fun setTextWatcher() {
        binding.searchEdt.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable) {}
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                applyQuery(s.toString())
            }
        })
        if (mCountryCodePicker.isKeyboardAutoPopOnSearch) {
            mInputMethodManager?.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0)
        }
    }

    /**
     * Filter country list for given keyWord / query.
     * Lists all countries that contains @param query in country's name, name code or phone code.
     *
     * @param query : text to match against country name, name code or phone code
     */
    private fun applyQuery(query: String) {
        var newQuery = query.lowercase(Locale.getDefault())
        binding.noResultTv.visibility = View.GONE

        //if query started from "+" ignore it
        if (newQuery.isNotEmpty() && newQuery[0] == '+') {
            newQuery = newQuery.substring(1)
        }
        mFilteredCountries = getFilteredCountries(newQuery)
        if (mFilteredCountries.isNullOrEmpty()) {
            binding.noResultTv.visibility = View.VISIBLE
        }
        mAdapter?.notifyDataSetChanged()
    }

    private fun getFilteredCountries(query: String): List<Country?> {
        val tempCountries: MutableList<Country?> = ArrayList()
        mCountryCodePicker.preferredCountries?.forEach { country ->
            if (country.isEligibleForQuery(query)) {
                tempCountries.add(country)
            }
        }
        if (tempCountries.size > 0) { //means at least one preferred country is added.
            tempCountries.add(null) // this will add separator for preference countries.
        }
        mCountryCodePicker.getCustomCountries()?.forEach { country->
            if (country.isEligibleForQuery(query)) {
                tempCountries.add(country)
            }
        }
        return tempCountries
    }

}